### PR TITLE
Add python type hints to parser.py

### PR DESCRIPTION
**Pull request by GeorgeSittas**
I made an attempt to add some type hints to the parser module. The "match" methods haven't been typed yet, because they cause a lot of errors. I may take another pass and see if I can type them too.

I've probably made some mistakes since there were a lot of changes involved, so please let me know if you find anything and I'll try to fix it.

On a side note: I'm not sure if the ROI is worth it for adding python types, in the end. It might be good for some other modules, but for the parser it didn't feel natural; returning the most general type "Expression" for most methods doesn't provide much information. Thoughts on this? (Besides that, it's definitely a frustrating process :-))

### CODE REVIEW

**Comment by tobymao** [comment]
@barakalon can you take a look and make a call?

File: sqlglot/parser.py [hunk 1]
```diff
@@ -681,7 +701,9 @@ def parse_into(self, expression_types, raw_tokens, sql=None):
             errors=merge_errors(errors),
         ) from errors[-1]
 
-    def _parse(self, parse_method, raw_tokens, sql=None):
+    def _parse(
+        self, parse_method: t.Callable, raw_tokens: t.List[Token], sql: t.Optional[str] = None
```
Comment (barakalon) [inline sqlglot/parser.py:705]: Should parse_method be `t.Callable[[Parser], Expression]`?

**Review by barakalon — COMMENTED** [review]
Looks good to me.

As for whether or not typing is worth it... 🤷‍♂️ 

_steam of consciousness..._

I think there are two groups to consider: **developers** of sqlglot and **users** of sqlglot.

For **users**, beyond annotating the main public methods, I think the main opportunity for type annotations would be on Expression args, as writing code that touches an AST is a bit fragile right now.

But given how Expression.arg_types works, that would be a _huge_ change. And it'd sacrifice parsing flexibility for type safety, which might not be worth it right now.

Annotating all these internal parse methods affects **developers** of sqlglot. 

Sqlglot has such a comprehensive test suite that I doubt this helps prevent many bugs. 

Does it help new folks contribute to the parser?

**Comment by GeorgeSittas** [comment]
Thanks for the feedback, Barak.

I agree with what you're saying. For me, the advantage of having type information is that it _sometimes_ makes the API a bit easier to understand and reason about (e.g. "what kind of arguments does this method expect? why?"). That is, I benefit by viewing type information as _documentation_.

Besides that, having to code in the correct types and/or iterate on an API to make it adapt to new changes sometimes adds an overhead to the whole development process, that would otherwise be avoided without having types altogether.

I'm interested to see how other people think about this topic too, though.

A related article I stumbled upon recently: https://lukeplant.me.uk/blog/posts/python-type-hints-parsy-case-study/

File: sqlglot/parser.py [hunk 2]
```diff
@@ -735,11 +765,26 @@ def raise_error(self, message, token=None):
...
         )
+
         if self.error_level == ErrorLevel.IMMEDIATE:
             raise error
+
         self.errors.append(error)
 
-    def expression(self, exp_class, comments=None, **kwargs):
+    def expression(
+        self, exp_class: exp._Expression, comments: t.Optional[t.List[str]] = None, **kwargs
```
[the remark this answers is not in the document]
Comment (GeorgeSittas) [reply sqlglot/parser.py]: Done.

File: sqlglot/parser.py (hunk 1 above)
Comment (GeorgeSittas) [reply sqlglot/parser.py]: Done

**Review by barakalon — APPROVED** [review]

**Comment by GeorgeSittas** [comment]
I'm checking the PR and after the cleanups I'll merge.

**Pull request merged**